code/parachain/frame/composable-maths/src/dex/constant_product.rs

```diff
@@ -91,6 +91,88 @@ where
 	Ok(ao)
 }
 
+/// Compute the amount of the output token given the amount of the input token.
+///
+/// If `Ok`, returns a tuple containing `(a_out, fee)`.
+/// To get `a_out` without accounting for the fee, set `f = 0`.
+/// Amount out, round down results.
+///
+/// **NOTE:** Weights must already be normalized.
+///
+/// # Parameters
+/// * `w_i` - Weight of the input token
+/// * `w_o` - Weight of the output token
+/// * `b_i` - Balance of the input token
+/// * `b_o` - Balance of the output token
+/// * `a_sent` - Amount of the input token sent by the user
+/// * `f` - Total swap fee
+///
+/// From https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#41-fee-math-updates,
+/// equation (2)
+pub fn compute_out_given_in_new<T: PerThing>(
+	w_i: T,
+	w_o: T,
+	b_i: u128,
+	b_o: u128,
+	a_sent: u128,
+	f: T,
+) -> Result<(u128, u128), ArithmeticError> {
+	let w_i = Decimal::from(w_i.deconstruct().into());
+	let w_o = Decimal::from(w_o.deconstruct().into());
+	let b_i = Decimal::from(b_i);
+	let b_o = Decimal::from(b_o);
+	let a_sent = Decimal::from(a_sent);
+
+	let weight_ratio = w_i.safe_div(&w_o)?;
+	// NOTE(connor): Use if to prevent pointless conversions if `f` is zero
+	let left_from_fee = if f.is_zero() {
+		Decimal::ONE
+	} else {
+		Decimal::from(f.left_from_one().deconstruct().into())
+			.safe_div(&Decimal::from(T::one().deconstruct().into()))?
+	};
+	let a_sent_fee_cut = a_sent.safe_mul(&left_from_fee)?;
+
+	let value = b_i.safe_add(&a_sent_fee_cut)?;
+	let value = b_i.safe_div(&value)?;
+	let value = value.checked_powd(weight_ratio).ok_or(ArithmeticError::Overflow)?;
+	let value = Decimal::ONE.safe_sub(&value)?;
```

Comment:
Same comment as the other PR - rename `value` here? This also seems to be a common-ish pattern, perhaps it could be extracted out somehow?

Reply:
Will go ahead and rename them, we can extract the functionality to a single function once all the code is present in the main. Otherwise, the refactoring and merge conflicts will be a pain.